### PR TITLE
Cache proto3 row encoding to reduce query consolidation memory util

### PR DESCRIPTION
**Pull request by bgwines**
## Background / Why?

When multiple clients consolidate on the same `SELECT`, the consolidator correctly shares a single `*sqltypes.Result` pointer across all waiters — zero-copy at the application layer. But each waiter then independently calls `ResultToProto3`, which calls `RowsToProto3`, copying all row byte data into new `querypb.Row.Values` buffers. For a 100MB result with 50 consolidated waiters, this alone produces ~5GB of transient allocations just from the row encoding step, before the gRPC codec even marshals the proto.

This change is inspired from hearing about a minor production event from the end of February, in which a cache expiration caused a burst of expensive queries. The consolidator correctly coalesced these repeated queries, but under high waiter counts the per-waiter memory cost — including redundant proto3 row encoding — contributed to a significant increase in VTTablet memory. Analysis indicated the consolidator was a significant contributor to that memory utilization.

With this change, the proto3 row encoding would have been computed once by the consolidator leader rather than N times by each waiter.

## Code Changes

- **New field** `proto3Rows` on `sqltypes.Result` caches the output of `RowsToProto3` for reuse across multiple `ResultToProto3` calls
- **`CacheProto3Rows()`** pre-computes and stores the encoding; nil-safe, no-op on empty results
- **`ResultToProto3`** checks for cached rows before falling through to the existing `RowsToProto3` path — zero behavior change for non-consolidated queries
- **Consolidator integration**: the leader calls `res.CacheProto3Rows()` before `Broadcast()`, so all waiters get pre-computed rows via the happens-before from the `RWMutex`
- **Feature flag**: `--consolidator-cache-proto3-rows` (default `false`) allows enabling the optimization at runtime without a redeploy

`go/sqltypes/cached_size.go` and `go/vt/proto/query/cached_size.go` are autogenerated by Sizegen (`make codegen`); the diff updates the struct size constant from 112 to ~136 to account for the new field. The `proto3Rows` cache is only populated on the non-OLAP consolidation path, so the OLAP consolidator's memory budgeting is unaffected

## Testing

- 4 new unit tests covering cached/uncached equivalence, nil/empty edge cases, `ShallowCopy` propagation, and `Copy` non-propagation
- 2 new cache invalidation tests verifying `AppendResult` and `Repair` nil out `proto3Rows`
- Ran full `go/sqltypes` and `go/vt/vttablet/tabletserver` test suites including race detector — all green

## Benchmarks

The benchmarks are in `go/sqltypes/proto3_test.go`. Here are the results from `go test -bench=BenchmarkResultToProto3 -benchmem -count=10` (per-call `ResultToProto3` cost, single consumer, varying result size):

```
┌─────────────┬───────────────────────┬───────────────────┬────────────────────┐
│             │        time/op        │    memory/op      │     allocs/op      │
├─────────────┼───────────────────────┼───────────────────┼────────────────────┤
│ Benchmark   │ uncached     │ cached │ uncached │ cached │ uncached  │ cached │
├─────────────┼──────────────┼────────┼──────────┼────────┼───────────┼────────┤
│ rows=100    │ 5.359µs ± 1% │ 23.50ns│ 19.77Ki  │ 144B   │ 302       │ 1      │
├─────────────┼──────────────┼────────┼──────────┼────────┼───────────┼────────┤
│ rows=1,000  │ 53.26µs ± 1% │ 25.24ns│ 195.6Ki  │ 144B   │ 3,002     │ 1      │
├─────────────┼──────────────┼────────┼──────────┼────────┼───────────┼────────┤
│ rows=10,000 │ 553.2µs ± 1% │ 23.50ns│ 1.909Mi  │ 144B   │ 30,002    │ 1      │
└─────────────┴──────────────┴────────┴──────────┴────────┴───────────┴────────┘
```

And here are the results from `go test -bench=BenchmarkConsolidationFanOut -benchmem -count=10` (10K-row result shared by _N_ concurrent goroutines, simulating the consolidator):

```
┌────────────┬────────────────────────────┬────────────────────┬────────────────────┐
│            │          time/op           │      memory/op     │      allocs/op     │
├────────────┼────────────────────────────┼────────────────────┼────────────────────┤
│ Benchmark  │ uncached     │  cached     │ uncached │ cached  │ uncached  │ cached │
├────────────┼──────────────┼─────────────┼──────────┼─────────┼───────────┼────────┤
│ waiters=1  │ 562.8µs ± 1% │ 323.4ns     │ 1.909Mi  │ 184B    │ 30,002    │ 3      │
├────────────┼──────────────┼─────────────┼──────────┼─────────┼───────────┼────────┤
│ waiters=10 │ 2.500ms      │ 2.157µs ± 1%│ 19.09Mi  │ 1.656Ki │ 300,031   │ 21     │
├────────────┼──────────────┼─────────────┼──────────┼─────────┼───────────┼────────┤
│ waiters=50 │ 12.40ms ± 1% │ 7.917µs     │ 95.47Mi  │ 8.219Ki │ 1,500,153 │ 101    │
└────────────┴──────────────┴─────────────┴──────────┴─────────┴───────────┴────────┘
```

---

### Pre-serialize the gRPC response

This PR eliminates the per-waiter proto3 row encoding, but a second N-fold copy remains one layer down. After `ResultToProto3` returns, each waiter independently marshals the `ExecuteResponse` to wire bytes via vtproto's `MarshalToSizedBufferVT`, which `copy()`s every `Row.Values` into a fresh serialization buffer. @venkatraju noted in verbal discussion that this would be much more effort (bypassing the per-message codec). I personally don't think it's worth it either, but I'm noting it as a possible secondary gain this PR doesn't provide.

### OLAP (streaming consolidator)

The same redundant-encoding problem exists in the streaming consolidator. However, the tradeoffs are different (e.g streaming chunks default to [32KB](https://github.com/slackhq/vitess/blob/18ff7c0b599a5ae9b5869614012e74a5f9bad5fb/go/vt/vttablet/tabletserver/tabletenv/config.go#L1096) (`StreamBufferSize`)). It may still be worthwhile under high fan-out, but needs separate analysis and measurement, and we seem to be skeptical of even having OLAP consolidation enabled at all: https://salesforce.enterprise.slack.com/archives/C0AKK6DKR71/p1775238112352109?thread_ts=1774527338501069&cid=C0AKK6DKR71

Co-Authored-By: Claude <svc-devxp-claude@slack-corp.com> but with every line reviewed by me :)

### CODE REVIEW

File: go/sqltypes/result.go [hunk 1]
```diff
@@ -37,6 +37,11 @@ type Result struct {
 	SessionStateChanges string           `json:"session_state_changes"`
 	StatusFlags         uint16           `json:"status_flags"`
 	Info                string           `json:"info"`
+
+	// proto3Rows caches the proto3-encoded representation of Rows, avoiding
+	// redundant encoding when multiple consumers share the same Result (e.g.,
+	// query consolidation where N clients share one query's result).
+	proto3Rows []*querypb.Row
```
Comment (bgwines) [inline go/sqltypes/result.go:45]: deliberately not JSON-expeorted since this is just a cached value. The proto3 representation wasn't part of the JSON before, no need to add it now.  It's also an implementation detail to the code so it shouldn't be exposed.

File: go/test/utils/diff.go [hunk 2]
```diff
@@ -78,7 +78,7 @@ func MustMatchFn(ignoredFields ...string) func(t *testing.T, want, got any, errM
 // Usage in Test*() function:
 //
 // testutils.MustMatch(t, want, got, "something doesn't match")
-var MustMatch = MustMatchFn()
+var MustMatch = MustMatchFn(".proto3Rows")
```
Comment (venkatraju) [inline go/test/utils/diff.go]: Adding `.proto3Rows` to the ignore list here feels too low level? I see `MustMatch` is used in a lot of places with many different kinds of protos compared.

File: go/flags/endtoend/vtcombo.txt [hunk 3]
```diff
@@ -47,6 +47,7 @@ Flags:
       --config-path strings                                              Paths to search for config files in. (default [{{ .Workdir }}])
       --config-persistence-min-interval duration                         minimum interval between persisting dynamic config changes back to disk (if no change has occurred, nothing is done). (default 1s)
       --config-type string                                               Config file type (omit to infer config type from file extension).
+      --consolidator-cache-proto3-rows                                   If true, the consolidation leader pre-caches proto3-encoded rows so that waiters avoid redundant encoding work.
```
Comment (venkatraju) [inline go/flags/endtoend/vtcombo.txt:50]: Is there a significant downside to enabling this behavior always? Wondering if we need the runtime control for this. Upstream may have a more informed opinion/preference - fine to keep this and ask when this is upstreamed.

**Review by venkatraju — COMMENTED** [review]
Looks great - should post this upstream. A benchmark for memory allocations in the consolidator with and without this change would be nice too.

File: go/flags/endtoend/vtcombo.txt (hunk 3 above)
Comment (bgwines) [reply go/flags/endtoend/vtcombo.txt:50]: thanks, I feel it should be default (and perhaps shouldn't even have a flag), but it seemed like a good idea to have the flag to roll it out safely. Curious as to how they'll feel regarding this matter (and how you feel given that context!)

**Comment by bgwines** [comment]
https://github.com/vitessio/vitess/pull/19872

**Comment by bgwines** [comment]
Closing in favor of #846 — the upstream PR (vitessio/vitess#19872) was merge-squashed, so the backport cherry-picks that single commit instead.
